code/implementation/docker/machine/src/main/java/io/cattle/platform/docker/machine/process/machinehost/MachineHostPreCreate.java

```diff
@@ -0,0 +1,65 @@
+package io.cattle.platform.docker.machine.process.machinehost;
+
+import static io.cattle.platform.core.model.tables.PhysicalHostTable.*;
+import static io.cattle.platform.docker.machine.constants.MachineHostConstants.*;
+import io.cattle.platform.core.model.PhysicalHost;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.object.util.DataUtils;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+import io.cattle.platform.util.type.Priority;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class MachineHostPreCreate extends AbstractObjectProcessLogic implements ProcessPreListener, Priority {
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { "physicalhost.create" };
+    }
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        PhysicalHost physHost = (PhysicalHost)state.getResource();
+
+        if ( !StringUtils.equalsIgnoreCase(physHost.getKind(), MACHINE_KIND) ) {
+            return null;
+        }
+
+        Map<Object, Object> newFields = new HashMap<Object, Object>();
+
+        if ( StringUtils.isEmpty(physHost.getExternalId()) ) {
+            String externalId = UUID.randomUUID().toString();
+            newFields.put(PHYSICAL_HOST.EXTERNAL_ID, externalId);
+        }
+
+        // TODO This sets a read-only "driver" field that will be useful to the
+        // UI.
+        // Should we set it here or in a request filter?
```

Comment:
Up to you, either is fine.  Being that the request filter checks these fields, I'd probably put it there.